### PR TITLE
8285671: java/nio/channels/etc/PrintSupportedOptions.java and java/nio/channels/DatagramChannel/AfterDisconnect.java are failing

### PR DESCRIPTION
**Pull request by Michael-Mc-Mahon**
Hi,

Can I get the following fix reviewed please? JDK-8284890 was pushed yesterday and is causing failures on older versions of macOS that do not support the option. The fix here is to check at initialization time whether it is supported before adding it to the list of supported options. The error causes the new test and two existing ones to fail. I will remove the two tests from the problem list separately.

Thanks,
Michael.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8285671](https://bugs.openjdk.java.net/browse/JDK-8285671): java/nio/channels/etc/PrintSupportedOptions.java and java/nio/channels/DatagramChannel/AfterDisconnect.java are failing

### Reviewers
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8419/head:pull/8419` \
`$ git checkout pull/8419`

Update a local copy of the PR: \
`$ git checkout pull/8419` \
`$ git pull https://git.openjdk.java.net/jdk pull/8419/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8419`

View PR using the GUI difftool: \
`$ git pr show -t 8419`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8419.diff">https://git.openjdk.java.net/jdk/pull/8419.diff</a>

</details>

### CODE REVIEW

**Comment by Michael-Mc-Mahon** [comment]
So, I had to remove the two problem listed tests, to start the review.

**Comment by Michael-Mc-Mahon** [comment]
Just spotted an issue with the native code check, if IPv6 is disabled, it will fail. The implementation should be able to work in that scenario.

File: test/jdk/jdk/net/ExtendedSocketOption/DontFragmentTest.java [hunk 1]
```diff
@@ -58,6 +58,9 @@ public static void main(String[] args) throws IOException {
     public static void testDatagramChannel() throws IOException {
         try (DatagramChannel c1 = DatagramChannel.open()) {
 
+            if (!c1.supportedOptions().contains(IP_DONTFRAGMENT)) {
+                return;
```
Comment (dfuch) [inline test/jdk/jdk/net/ExtendedSocketOption/DontFragmentTest.java:62]: Should we assert that this should only happens on mac here? Same remark for the two other places below...

Comment (Michael-Mc-Mahon) [reply test/jdk/jdk/net/ExtendedSocketOption/DontFragmentTest.java:62]: I think that is reasonable.

File: test/jdk/jdk/net/ExtendedSocketOption/DontFragmentTest.java [hunk 2]
```diff
@@ -38,7 +38,10 @@
 
 public class DontFragmentTest {
 
+    private static boolean isMacos;
+
     public static void main(String[] args) throws IOException {
+        isMacos = System.getProperty("os.name").equals("Mac OS X");
```
Comment (dfuch) [inline test/jdk/jdk/net/ExtendedSocketOption/DontFragmentTest.java:44]: I believe there's a test library class that does that. I never remember what the os.name is supposed to look like.

File: src/jdk.net/macosx/native/libextnet/MacOSXSocketOptions.c [hunk 3]
```diff
@@ -189,6 +189,35 @@ static int socketFamily(jint fd) {
...
+    fd = socket(AF_INET, SOCK_DGRAM, 0);
+    if (fd == -1)
+        return JNI_FALSE;
+    value = 1;
+    rv = setsockopt(fd, IPPROTO_IP, IP_DONTFRAG, &value, sizeof(value));
+    close(fd);
+    if (rv == -1) {
+        return JNI_FALSE;
+    }
+    fd = socket(AF_INET6, SOCK_DGRAM, 0);
```
[the remark this answers is not in the document]
Comment (Michael-Mc-Mahon) [reply src/jdk.net/macosx/native/libextnet/MacOSXSocketOptions.c:209]: I'm not sure you can disable IPv6 completely on mac OS. You can disable it on a per-interface basis. Even then it leaves a link local address available and that code succeeds.

File: test/jdk/jdk/net/ExtendedSocketOption/DontFragmentTest.java (hunk 2 above)
Comment (Michael-Mc-Mahon) [reply test/jdk/jdk/net/ExtendedSocketOption/DontFragmentTest.java]: Okay. It's `Platform.isOSX()`

File: src/jdk.net/macosx/native/libextnet/MacOSXSocketOptions.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (dfuch) [reply src/jdk.net/macosx/native/libextnet/MacOSXSocketOptions.c:209]: OK - if you're confident that this will work.

**Review by dfuch — APPROVED** [review]

**Pull request closed without merging**